### PR TITLE
I745

### PR DESCRIPTION
**Pull request by littledice**
* New `OrderlyServer` class to make requests to orderly server.
* Add new functions to `ActionContext` to get post data from the request and set status code of the reponse.
* 7 new routes using the above
* Split out routing into separate classes for readiability
* Updated `generateTestData` task and created `startOrderlyServer` task to support integration testing

### CODE REVIEW

File: src/app/src/main/kotlin/org/vaccineimpact/reporting_api/Helpers.kt [hunk 1]
```diff
@@ -23,9 +23,12 @@ fun addTrailingSlashes(req: Request, res: Response)
 
 fun addDefaultResponseHeaders(res: HttpServletResponse, contentType: String)
 {
-    res.contentType = contentType
-    res.addHeader("Content-Encoding", "gzip")
-    res.addHeader("Access-Control-Allow-Origin", "*")
+    if (!res.containsHeader("Content-Encoding"))
+    {
+        res.contentType = contentType
```
Comment (littledice) [inline src/app/src/main/kotlin/org/vaccineimpact/reporting_api/Helpers.kt]: This `if` statement feels a bit hack-y but not sure of a better way around this issue:

Spark's `after` filter is matching some routes multiple times and so trying to apply this function multiple times to the same response.

This happens on `/reports/:name/run/` because it matches both its intended route and also `/reports/:name/:version/`

**Comment by littledice** [comment]
This is still a lot of file changes! Let me know if you want me to split it out further or to go through it in person tomorrow.

File: scripts/build-app.sh [hunk 2]
```diff
@@ -13,9 +13,18 @@ docker build --tag montagu-reporting-api-app-build \
     -f app.Dockerfile \
 	.
 
+docker pull docker.montagu.dide.ic.ac.uk:5000/orderly.server:master
+
+docker run --rm \
+    -p 8123:8123 \
+    -v $PWD/git:/orderly \
+    --network=host \
+    docker.montagu.dide.ic.ac.uk:5000/orderly.server:master "orderly" &
```
Comment (MartinEden) [inline scripts/build-app.sh]: You probably want to use Docker's `-d` (detached) flag rather than `&` to make this run in the background.

File: spec/test/test-schema.py [hunk 3]
```diff
@@ -1,84 +0,0 @@
...
-            example = get_example(lines)
-            if example:
-                validate(url, example)        
-            else:
-                raise Exception("No example given for {}".format(url))
-        
-        line = get_next(lines)
-
-check_spec(spec_path)
-print("Finished without errors ☺")
```
Comment (MartinEden) [inline spec/test/test-schema.py]: Feels like these spec related removals should maybe be in another PR?

File: src/app/build.gradle [hunk 4]
```diff
@@ -84,18 +85,32 @@ task generateTestData() {
...
             commandLine "rm", "$projectDir/demo", "--recursive"
         }
+        exec {
+            commandLine "rm", "$projectDir/git", "--recursive"
+        }
+    }
+}
+
+task startOrderlyServer() {
+    String image = "docker.montagu.dide.ic.ac.uk:5000/orderly.server:i648"
```
Comment (MartinEden) [inline src/app/build.gradle]: "i648" should be lifted out into a config file.

File: scripts/build-app.sh [hunk 5]
```diff
@@ -13,9 +13,18 @@ docker build --tag montagu-reporting-api-app-build \
     -f app.Dockerfile \
 	.
 
+docker pull docker.montagu.dide.ic.ac.uk:5000/orderly.server:master
```
Comment (MartinEden) [inline scripts/build-app.sh]: Should probably use a configurable version, rather than master?

File: scripts/start-orderly-server.sh [hunk 6]
```diff
@@ -0,0 +1,8 @@
+IMAGE=docker.montagu.dide.ic.ac.uk:5000/orderly.server:i648
```
Comment (MartinEden) [inline scripts/start-orderly-server.sh]: "i648" should be lifted out into a config file.

File: src/app/src/main/kotlin/org/vaccineimpact/reporting_api/DirectActionContext.kt [hunk 7]
```diff
@@ -49,4 +54,19 @@ open class DirectActionContext(private val context: SparkWebContext) : ActionCon
...
+
+    override fun setStatusCode(statusCode: Int)
+    {
+        response.status(statusCode)
+    }
+    override fun postData(): Map<String, String>
+    {
+        val body = request.body()
+
+        if (body.isNullOrEmpty())
```
Comment (MartinEden) [inline src/app/src/main/kotlin/org/vaccineimpact/reporting_api/DirectActionContext.kt]: I prefer to be explicit about `if` braces, as well as `else` statements when both branches return. This is a stylistic point that you're free to disagree with. But I think it's clearer like this:

```
return if (body.isNullOrEmpty())
{
    emptyMap()
}
else
{
    GsonBuilder().create().fromJson<Map<String, String>>(request.body())
}

File: src/app/src/main/kotlin/org/vaccineimpact/reporting_api/Helpers.kt (hunk 1 above)
Comment (MartinEden) [reply src/app/src/main/kotlin/org/vaccineimpact/reporting_api/Helpers.kt]: I think the way to go with this is to make the `addDefaultResponseHeaders` properly idempotent. Like so:

```
res.contentType = contentType // I'm assuming that since there is a property for this, Spark handles only emitting one header correctly?
if (!res.containsHeader("Content-Encoding"))
{
    res.addHeader("Content-Encoding", "gzip")
}
if (!res.containsHeader("Access-Control-Allow-Origin"))
{
    res.addHeader("Access-Control-Allow-Origin", "*")
}
```

However, I would *also* encourage that we don't have ambiguous URLs, ever. Could you change the URL scheme to:
* `/reports/:name/run/`
* `/reports/:name/versions/:version/`

Then it's unambiguous.

File: src/app/src/main/kotlin/org/vaccineimpact/reporting_api/OrderlyServerAPI.kt [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+        return httpClient.get(fullUrl, standardHeaders)
+    }
+
+    private fun buildFullUrl(url: String, queryString: String?): String
+    {
+        if (queryString == null)
+            return "$urlBase$url"
+
+        return "$urlBase$url?$queryString"
+    }
```
Comment (MartinEden) [inline src/app/src/main/kotlin/org/vaccineimpact/reporting_api/OrderlyServerAPI.kt]: Maybe rewrite this as:

```
val queryPart = if (queryString != null)
{
    "?" + queryString 
}
else
{
    ""
}
return urlBase + url + queryPart
```

File: src/app/src/main/kotlin/org/vaccineimpact/reporting_api/controllers/GitController.kt [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+                    private val orderlyServerAPI: OrderlyServerAPI) : Controller(context)
+{
+
+    constructor(context: ActionContext) :
+            this(context, OrderlyServer(AppConfig(), KHttpClient()))
+
+    fun status(): String
+    {
+        val response = orderlyServerAPI.get("/reports/git/status/", context)
+        return returnResponse(response)
```
Comment (MartinEden) [inline src/app/src/main/kotlin/org/vaccineimpact/reporting_api/controllers/GitController.kt]: `return returnResponse(response)` is a little confusing. :)

Maybe rename to `passThroughResponse`, or something similar?

Comment (littledice) [reply src/app/src/main/kotlin/org/vaccineimpact/reporting_api/controllers/GitController.kt]: ah yes, this was another point I meant to ask for advice on!

File: src/app/src/main/kotlin/org/vaccineimpact/reporting_api/Helpers.kt (hunk 1 above)
Comment (littledice) [reply src/app/src/main/kotlin/org/vaccineimpact/reporting_api/Helpers.kt]: Yeah, I did consider this, but a little reluctant to change the endpoints now (I'd also then want to change all the `/report/:name/:version/artefacts/:artefact` type endpoints for consistency.) Maybe we could change the new one instead; would `/reports/run/:name` work?

File: scripts/start-orderly-server.sh (hunk 6 above)
Comment (littledice) [reply scripts/start-orderly-server.sh]: Think this whole file is actually obsolete. Will remove.

File: spec/test/test-schema.py (hunk 3 above)
Comment (littledice) [reply spec/test/test-schema.py]: https://github.com/vimc/montagu-reporting-api/pull/17/

File: src/app/src/main/kotlin/org/vaccineimpact/reporting_api/Helpers.kt (hunk 1 above)
Comment (richfitz) [reply src/app/src/main/kotlin/org/vaccineimpact/reporting_api/Helpers.kt]: Let me know what gets changed as I should reflect this in the underlying orderly server code for consistency

**Comment by littledice** [comment]
Ok, 
* extra PR for the file deletions: https://github.com/vimc/montagu-reporting-api/pull/17
* made suggested code style changes
* versioned orderly server
* removed unused shell script
* see comment on routing

**Comment by MartinEden** [comment]
Two more things I'd like you to do:
* Add that `tests.finalizedBy` block we discussed
* Open a ticket for rearranging the endpoints to be non-ambiguous. I agree that's probably out of scope here (and it would be good to get this out the door). But I think we should still try and fix the endpoint ambiguity now, when no one's using the API directly - it will get harder later.

**Comment by MartinEden** [comment]
Then you can merge this without waiting for review from me (assuming TeamCity stays green!)

**Comment by littledice** [comment]
I've actually taken both those `test.dependsOn` and `test.finalizedBy` rules out of the project; users will have to manually start and stop the orderly server (because I couldn't get it to work the same way in docker.) I'll update the README accordingly.

**Comment by littledice** [comment]
Opened: https://vimc.myjetbrains.com/youtrack/issue/VIMC-799 (resolve routing ambiguity)
